code/datums/alert.dm

```diff
@@ -0,0 +1,181 @@
+///Important to note is that source_atom is not held as a ref, we're used as a proxy to prevent hard deletes
+///optional_camera should only be used when you have one camera you want to pass along to alert listeners, most of the time you should have no use for it
+/datum/alert_handler/proc/send_alert(alert_type, atom/use_as_source_atom = source_atom, optional_camera)
+	if(!use_as_source_atom)
+		return
+
+	var/area/our_area = get_area(use_as_source_atom)
+	var/our_z_level = use_as_source_atom.z
+
+	if (our_area.area_flags & NO_ALERTS)
+		return FALSE
+
+	var/list/existing_alerts = sent_alerts[alert_type]
+	if(existing_alerts)
+		if(our_area in existing_alerts)
+			return FALSE
+	else
+		sent_alerts[alert_type] = list()
+		existing_alerts = sent_alerts[alert_type]
+
+	existing_alerts += our_area
+
+	our_area.active_alarms[alert_type] += 1
+
+	SEND_GLOBAL_SIGNAL(COMSIG_ALERT_FIRE(alert_type), src, alert_type, our_area, our_z_level, optional_camera)
+
+	return TRUE
+
+///Clears an alert from any interested listeners
+/datum/alert_handler/proc/clear_alert(alert_type, use_as_source_atom = source_atom)
+	if(!use_as_source_atom)
+		return
+
+	var/area/our_area = get_area(use_as_source_atom)
+	return clear_alert_from_area(alert_type, our_area)
+
+///Exists so we can request that the alerts from an area are cleared, even if our source atom is no longer in that area
+/datum/alert_handler/proc/clear_alert_from_area(alert_type, area/our_area)
+	if (our_area.area_flags & NO_ALERTS)
+		return FALSE
+
+	var/list/existing_alerts = sent_alerts[alert_type]
+	if(!existing_alerts)
+		return FALSE
+
+	if(!(our_area in existing_alerts))
+		return FALSE
+
+	existing_alerts -= our_area
+	if(!length(existing_alerts))
+		sent_alerts -= alert_type
+
+	our_area.active_alarms[alert_type] -= 1
+	if(!length(our_area.active_alarms))
+		our_area.active_alarms -= alert_type
+
+	SEND_GLOBAL_SIGNAL(COMSIG_ALERT_CLEAR(alert_type), src, alert_type, our_area)
+	return TRUE
+
+/datum/alert_listener
+	///List of valid source z levels, ignored if null
+	var/list/allowed_z_levels
+	///List of allowed areas. if this is null it's ignored
+	var/list/allowed_areas
+
+	///List of alert type -> list of area name -> list(area, ref to area's cameras, list(sources))
+	var/list/alarms = list()
+	///Should we allow alert changes to go through or not
+	var/accepting_alert_changes = TRUE
+
+///Accepts a list of alert types to pay attention to, a list of valid z levels, and a list of valid areas. areas and zlevels are ignored if null
+/datum/alert_listener/New(_alerts_to_listen_for, _allowed_z_levels, _allowed_areas)
+	allowed_z_levels = _allowed_z_levels
+	allowed_areas = _allowed_areas
+	for(var/alert_type in _alerts_to_listen_for)
+		RegisterSignal(SSdcs, COMSIG_ALERT_FIRE(alert_type), .proc/add_alert)
+		RegisterSignal(SSdcs, COMSIG_ALERT_CLEAR(alert_type), .proc/clear_alert)
+
+	return ..()
+
+///Adds an alert to our alarms list, you probobly shouldn't be calling this manually
```

Comment:
```suggestion
///Adds an alert to our alarms list, you probably shouldn't be calling this manually
```

Can you also describe when you would want to call this manually, and what you'd want to use instead?

Reply:
yeah sure